### PR TITLE
WIP: 7 add timeout type to api

### PR DESCRIPTION
**Pull request by paweljaciow**

### CODE REVIEW

File: proto/dictation_asr.proto [hunk 1]
```diff
@@ -418,46 +506,110 @@ message StreamingRecognizeResponse {
...
+
+    // [**Extension by Techmo**]
+    // This event indicates that recognition has started and there was no speech detected
+    // for a certain period of time. The server will not process
+    // additional audio (although it may subsequently return additional results).
+    // The client should stop sending additional audio data, half-close the gRPC
+    // connection, and wait for any additional results until the server closes
+    // the gRPC connection.
+    // MRCPv2-related: [No-Input-Timeout](https://tools.ietf.org/html/rfc6787#section-9.4.6)
+    NO_INPUT_TIMEOUT = 2;
```
Comment (tomjad) [inline proto/dictation_asr.proto]: Probably we should remove 'Timeout' from all fields in this `SpeechEventType` enum. `SpeechEvent::NO_INPUT` sounds better than `NO_INPUT_TIMEOUT`.

File: proto/dictation_asr.proto [hunk 2]
```diff
@@ -418,46 +506,110 @@ message StreamingRecognizeResponse {
...
+
     // [**Extension by Techmo**]
-    // This event indicates that the server has detected no speech input timeout
-    // and expects no additional speech. Therefore, the server will not process
+    // This event indicates that recognition has started and there was no result
```
Comment (tomjad) [inline proto/dictation_asr.proto]: This is not true. We still can return results that were processed up to the time when `Recognition Timeout` occured.

File: proto/dictation_asr.proto [hunk 3]
```diff
@@ -418,46 +506,110 @@ message StreamingRecognizeResponse {
...
+    // for a certain period of time. The server will not process
     // additional audio (although it may subsequently return additional results).
     // The client should stop sending additional audio data, half-close the gRPC
     // connection, and wait for any additional results until the server closes
     // the gRPC connection.
-    NO_INPUT_TIMEOUT = 4;
+    // MRCPv2-related: [Recognition-Timeout](https://tools.ietf.org/html/rfc6787#section-9.4.7)
+    RECOGNITION_TIMEOUT = 3;
```
Comment (tomjad) [inline proto/dictation_asr.proto]: `Can you consider **TOO_MUCH_SPEECH** as a name for this Event ?

File: proto/dictation_asr.proto [hunk 4]
```diff
@@ -418,46 +506,110 @@ message StreamingRecognizeResponse {
...
+
+    // [**Extension by Techmo**]
+    // [Speech-Complete-Timeout](https://tools.ietf.org/html/rfc6787#section-9.4.15)
+    // This event indicates that recognition has been finalized with a complete result
+    // after specified length of silence after user speech.
+    SPEECH_COMPLETE_TIMEOUT = 4;
```
Comment (tomjad) [inline proto/dictation_asr.proto]: Remove TIMEOUT, this is standard case. Do we assume that from now we're returning Event type for every message ?

File: proto/dictation_asr.proto [hunk 5]
```diff
@@ -418,46 +506,110 @@ message StreamingRecognizeResponse {
...
+
+    // [**Extension by Techmo**]
+    // [Speech-Incomplete-Timeout](https://tools.ietf.org/html/rfc6787#section-9.4.16)
+    // This event indicates that recognition has been finalized with an incomplete result
+    // after specified length of silence after user speech.
+    SPEECH_INCOMPLETE_TIMEOUT = 5;
```
Comment (tomjad) [inline proto/dictation_asr.proto]: Remove TIMEOUT

File: proto/dictation_asr.proto [hunk 6]
```diff
@@ -418,46 +506,110 @@ message StreamingRecognizeResponse {
...
   }
 
-  // [*Output-only*] If set, returns a [google.rpc.Status][google.rpc.Status] message that
+  // [**Extension by Techmo**]
+  // Indicates the type of speech processing timeout. These are MRCPv2-related.
+//  enum SpeechProcessingTimeoutType {
```
Comment (tomjad) [inline proto/dictation_asr.proto]: Why it's commented here?

Comment (paweljaciow) [reply proto/dictation_asr.proto]: I did not clean up after my previous idea. I will remove it.

File: proto/dictation_asr.proto (hunk 4 above)
Comment (paweljaciow) [reply proto/dictation_asr.proto]: It looks like we do. For every result, an event indicating recognition termination cause will be sent first. Do you see any potential drawback of this approach?

Comment (tomjad) [reply proto/dictation_asr.proto]: Can we simplify communication flow and return both event type and results (when applicable) in single message?

File: proto/dictation_asr.proto (hunk 2 above)
Comment (paweljaciow) [reply proto/dictation_asr.proto]: Based on [this](https://tools.ietf.org/html/rfc6787#section-9.4.7), I will change it to:
```
...
This event indicates that recognition has started but has not finished
due to too much speech has been received for a certain period of time.
The server will not process ...
...
```
Does it sound correct?

File: proto/dictation_asr.proto (hunk 4 above)
Comment (paweljaciow) [reply proto/dictation_asr.proto]: I would say no. From GSAPI:
```cpp
// - In each response, only one of these fields will be set:
//     `error`,
//     `speech_event_type`, or
//     one or more (repeated) `results`.
```

File: proto/dictation_asr.proto [hunk 7]
```diff
@@ -507,74 +507,14 @@ message StreamingRecognizeResponse {
...
-    // [Speech-Incomplete-Timeout](https://tools.ietf.org/html/rfc6787#section-9.4.16)
-    // This event indicates that recognition has been finalized with an incomplete result
-    // after specified length of silence after user speech.
-    SPEECH_INCOMPLETE_TIMEOUT = 5;
+    // This event indicates that the server expects no additional speech
+    // due to audio processing termination (although it may subsequently return
+    // additional results). The client should stop sending additional audio data,
+    // half-close the gRPC connection, and wait for any additional results
+    // until the server closes the gRPC connection.
+    STOPPED_ACCEPTING_AUDIO = 2;
```
[the remark this answers is not in the document]
Comment (paweljaciow) [reply proto/dictation_asr.proto]: In a private discussion we decided to:  
- leave only `END_OF_SINGLE_UTTERANCE` event as it is in the original API,  
- change handling no input timeout to be unified with other timeout types that will result in using only `END_OF_SINGLE_UTTERANCE` to signalize stream closure to the client,
- not include timeouts in synchronous recognition.

File: proto/dictation_asr.proto [hunk 8]
```diff
@@ -587,15 +527,38 @@ message StreamingRecognizeResponse {
...
-  // [*Output only*] Indicates the type of speech processing timeout.
-//  SpeechProcessingTimeoutType speech_processing_timeout_type = 6;
 }
 
 // A streaming speech recognition result corresponding to a portion of the audio
 // that is currently being processed.
 message StreamingRecognitionResult {
+  // [**Extension by Techmo**]
+  // Indicates the cause of audio processing termination. These are MRCPv2-related.
+  enum AudioProcessingTerminationCause {
```
[the remark this answers is not in the document]
Comment (paweljaciow) [reply proto/dictation_asr.proto]: Done.

**Pull request merged**